File: leetcode/extension-problems/1.two-sum.go
/*
 * @lc app=leetcode id=1 lang=golang
 *
 * [1] Two Sum
 */

// @lc code=start
func twoSum(nums []int, target int) []int {
	res := []int{-1, -1}
	if len(nums) < 2 {
		return res
	}

	hashMap := map[int]int{}
	other := 0
	for i := range nums {
		other = target - nums[i]
		if _, ok := hashMap[other]; ok {
			res[0] = hashMap[other]
			res[1] = i
		}

		hashMap[nums[i]] = i
	}

	return res
}

// @lc code=end

File: leetcode/problems/sessions/2023-05-01_2023-??-??/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#


# @lc code=start

# * Hash Table Solution | O(n) Time | O(n) Space
# * n -> The length of nums array


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_idx_map = {}
        for idx, num in enumerate(nums):
            other_val = target - num
            if other_val in num_idx_map:
                return [num_idx_map[other_val], idx]

            num_idx_map[num] = idx


# @lc code=end
